### ISSUE TITLE
Issue with MJ's handling of gimbalResponseSpeed

### ISSUE THREAD

**Issue by Starwaster** [MuMech/MechJeb2#1859]
Looking at how MJ factors in gimbalResponseSpeed into torqueReactionSpeed6

torqueReactionSpeed6.Add(Mathf.Abs(g.gimbalRange) / g.gimbalResponseSpeed * Vector3d.Max(pos.Abs(), neg.Abs()));

from

https://github.com/MuMech/MechJeb2/blob/7ebb844a5363ae000792710b1756dc03a2f3bc23/MechJeb2/VesselState.cs#L1081

This code is making wrong assumptions that are compounded by bad stock code. The first problem is that gimbalRange is irrelevant to how long it takes the gimbal to reach maximum deflection. It actually takes the same amount of time no matter how far it has to gimbal.

ModuleGimbal does: lerp(current pos, goal pos, gimbalResponseSpeed * fixedDeltaTime)

Because of the way lerp works, the default speed of 10 means that gimbaling happens in 5 frames (1/10th of a second)
It also means that the gimbal rate is not constant. Highest rate is on the first frame. Assuming gimbal range of 10, that looks something like:

Frame 1 = 2		(100 deg/sec)
Frame 2 = 5.2		(80 deg/sec)
Frame 3 = 8.08	(48 deg/sec)
Frame 4 = 9.616	(19.2 deg/sec)
Frame 5 = 10		(3.84 deg/sec)

Removing gimbalRange from MJ's handling will help, but I'm not sure what the answer is to ModuleGimbal's implementation. lerp shouldn't be used the way they're using it. (using the current value to lerp **FROM** guarantees the rate can't be constant and multiplying a clamped value by speed means it's converging on its goal a lot sooner than we think it is)

**Comment by lamont-granquist** [comment]
I don't see a lot in MJ that depends on the torqueReactionSpeed other than the old MJ PID, but this is definitely useful information to have about how gimbal response works in KSP since there's some deficiencies in the "bettercontroller" that i think could get fixed by adjustments to take into account gimbal repsonse (which it entirely ignores).  By knowing what KSP does I can probably simulate the response speed delays in matlab/simulink and get some PID scheduling out of it.  Unfortunately, I'm hacking on PVG stuff right now and burned out on PID things awhile back and need some time to pass before I start wanting to scratch that itch again.  This is definitely good research though.

**Comment by lamont-granquist** [comment]
Everything anyone would want to know about thrust vector control and how to model it and control it:

https://ntrs.nasa.gov/api/citations/20230000427/downloads/AAS_2023_P3_Orr_TVC_Modeling_paper.pdf

**Comment by lamont-granquist** [comment]
So doesn't KSP "re-lerp" on every frame?  So what I get for slewing from 0 to 10 at a fixedDeltaTime of 0.02 and gimbalResponseSpeed of 10 looks like (zeno's paradox, always getting 20% closer each step):

```
0.00: 0.000000
0.02: 2.000000
0.04: 3.600000
0.06: 4.880000
0.08: 5.904000
0.10: 6.723200
0.12: 7.378560
0.14: 7.902848
0.16: 8.322278
0.18: 8.657823
0.20: 8.926258
0.22: 9.141007
0.24: 9.312805
0.26: 9.450244
0.28: 9.560195
0.30: 9.648156
0.32: 9.718525
0.34: 9.774820
0.36: 9.819856
0.38: 9.855885
0.40: 9.884708
0.42: 9.907766
```

While if fixedDeltaTime is cranked up to 0.08 (4x phys warp) then it looks like (getting 80% closer each step):

```
0.00: 0.000000
0.08: 8.000000
0.16: 9.600000
0.24: 9.920000
```

So that looks like a P-controller (pid) with a wonky incorrect adjustment for warp, and explains why sometimes gimbal wobbles go away at higher warp.

**Comment by lamont-granquist** [comment]
no, i'm very wrong, that is a discrete low pass filter, simple first order IIR filter:  

```
y(k+1) = 0.8*y(k) + 0.2*r(k)
```

or:

```
G(z) = 0.2/(z - 0.8)
```

or:

```
Ts = 0.02;
num = 0.2;
den = [1 -0.8];
sys = tf(num, den, Ts);
```

Using d2c to convert that to a continuous time filter results in:

```
G(s) = 11.16 / (s + 11.16)
```

 Converting that back to discrete with a timestep of 0.08 results in (how phys timewarp _should_ behave):

```
G(z) = 0.5904 / (z - 0.4096)
```

Compare to the actual KSP code which results in:

```
tf(0.80, [1 -0.20], 0.08)
G(z) = 0.80 / (z - 0.20)
G(s) = 20.12 / (s + 20.12)
```

The cutoff frequency just about doubles, making the engine gimbal more responsive.

**Comment by lamont-granquist** [comment]
The BetterController implementation now just exploits the fact that the KSP gimbal implementation is a low pass filter.
